frappe/email/doctype/notification/notification.js

```diff
@@ -112,5 +108,30 @@ frappe.ui.form.on("Notification", {
 	},
 	channel: function(frm) {
 		frm.toggle_reqd("recipients", frm.doc.channel=="Email");
+	},
+	subject: function(frm) {
+		last_touched = 'subject';
+	},
+	message: function(frm) {
+		last_touched = 'message';
+	},
+	insert_dynamic_value: function(frm) {
+		var dialog = new frappe.ui.Dialog({
+			title: __('Insert Dynamic Value'),
+			fields: [
+				{
+					fieldtype: 'Autocomplete',
+					fieldname: 'fieldname',
+					label: __('Select Field'),
+					options: options,
+					reqd: 1,
+					onchange: function() {
+						frm.set_value(last_touched, `${frm.doc[last_touched]} {{ doc.${this.value} }} `);
```

Comment:
With this users will not be able to add the syntax between two strings. It will always append to the end.

You might have to remember user's `last_cursor_position` in the `last_focused_field` and add the dynamic field syntax accordingly.